### PR TITLE
Feature/multiple adapter support

### PR DESCRIPTION
**Pull request by kasvith**
Pull Request for Issue #190  .

### Summary of Changes
- Added config param to Local File System plugin for set Display Name
- Added Foo File System Plugin
- Added multiple adapter support to API
- Added multiple adapter support to Controller
- Injected config params to Media View

### Testing Instructions
- Discover **Foo Filesystem**
- Go to NMM
- HTTP 200OK message will be shown, ignore for now
- Try these urls on browser
  - http://localhost/media-manager-improvement/administrator/index.php?option=com_media&format=json&task=api.files&path=local:/banners
  - http://localhost/media-manager-improvement/administrator/index.php?option=com_media&format=json&task=api.files&path=foo:/

### Expected result
- A JSON response containing all the files

### Actual result

@dneukirchen @dgt41 if this is ok we can go with frontend implementation

### CODE REVIEW

File: administrator/components/com_media/models/api.php [hunk 1]
```diff
@@ -38,85 +46,128 @@ public function __construct($config = array())
...
+			// Import enabled file system plugins
 			JPluginHelper::importPlugin('filesystem');
 
 			$app = JFactory::getApplication();
 
 			$results = $app->triggerEvent('onFileSystemGetAdapters');
+			$adapters = array();
 
-			if ($results != null)
+			for ($i = 0; $i < count($results); $i++)
```
Comment (dgrammatiko) [inline administrator/components/com_media/models/api.php]: `for ($i = 0, $l = count($results); $i < $l; $i++)` is faster

**Review by dgrammatiko — APPROVED** [review]

File: administrator/components/com_media/models/api.php (hunk 1 above)
Comment (kasvith) [reply administrator/components/com_media/models/api.php]: but count($results) counted when loop initialize right?

Comment (kasvith) [reply administrator/components/com_media/models/api.php]: @dgt41 if $results = null, count($results) would return a 0, so loop will terminate at that point

File: administrator/components/com_media/controllers/api.json.php [hunk 2]
```diff
@@ -81,7 +81,17 @@ class MediaControllerApi extends Controller
 	public function files()
 	{
 		// Get the required variables
-		$path = $this->input->getPath('path', '/', 'path');
+		$pathInfo = explode(':', $this->input->getString('path', 'local:/'));
+		$adapterInfo = explode('-', $pathInfo[0]);
+		$path = $pathInfo[1];
+		$adapter = $adapterInfo[0];
+		$accountID = "";
```
Comment (yvesh) [inline administrator/components/com_media/controllers/api.json.php]: Codestyle - please indent all = on the same

```
$pathInfo = ..
$path       = ..
```

File: administrator/components/com_media/controllers/api.json.php [hunk 3]
```diff
@@ -81,7 +81,17 @@ class MediaControllerApi extends Controller
 	public function files()
 	{
 		// Get the required variables
-		$path = $this->input->getPath('path', '/', 'path');
+		$pathInfo = explode(':', $this->input->getString('path', 'local:/'));
+		$adapterInfo = explode('-', $pathInfo[0]);
+		$path = $pathInfo[1];
+		$adapter = $adapterInfo[0];
```
Comment (yvesh) [inline administrator/components/com_media/controllers/api.json.php]: you can use php function list instead :-)

File: administrator/components/com_media/controllers/api.json.php (hunk 2 above)
Comment (kasvith) [reply administrator/components/com_media/controllers/api.json.php]: sure

File: administrator/components/com_media/models/api.php (hunk 1 above)
Comment (dgrammatiko) [reply administrator/components/com_media/models/api.php]: if it's in the second part of the for then count() is calculated in every iteration ;)

File: administrator/components/com_media/models/api.php [hunk 4]
```diff
@@ -38,85 +46,128 @@ public function __construct($config = array())
...
+	 * @return MediaFileAdapterInterface
+	 *
+	 * @throws Exception
+	 */
+	private function getAdapter($name)
+	{
+		if (isset($this->adapters[$name]))
+		{
+			return $this->adapters[$name];
+		}
```
Comment (yvesh) [inline administrator/components/com_media/models/api.php]: You don't need else here, as you do a return ;-)

File: administrator/components/com_media/models/api.php [hunk 5]
```diff
@@ -38,85 +46,128 @@ public function __construct($config = array())
...
+		else
 		{
-			$this->adapter = $config['fileadapter'];
+			// Todo Use a translated string
+			throw new Exception('Requested media file adapter was not found', 300);
```
Comment (yvesh) [inline administrator/components/com_media/models/api.php]: Why 300? :D

File: administrator/components/com_media/views/media/tmpl/default.php [hunk 6]
```diff
@@ -11,6 +11,30 @@
...
+ */
+function getProviderInfo()
+{
+	$providerInfo = JPluginHelper::getPlugin('filesystem');
+	$results = array();
+
+	foreach ($providerInfo as $provider)
+	{
+		$params = new JRegistry($provider->params);
+		$info = new stdClass;
```
Comment (yvesh) [inline administrator/components/com_media/views/media/tmpl/default.php]: Indention on the longest equal sign :-)

File: administrator/components/com_media/models/api.php (hunk 5 above)
Comment (kasvith) [reply administrator/components/com_media/models/api.php]: anything better 🥇

File: administrator/components/com_media/views/media/tmpl/default.php [hunk 7]
```diff
@@ -11,6 +11,30 @@
...
 $params = JComponentHelper::getParams('com_media');
 
+/**
+ * Obtain list of supported providers
+ *
+ * @return string
+ *
+ * @since __DEPLOY_VERSION__
+ */
+function getProviderInfo()
```
Comment (laoneo) [inline administrator/components/com_media/views/media/tmpl/default.php]: This needs to be done in the view, the template files should not have functions.

File: plugins/filesystem/local/local.xml [hunk 8]
```diff
@@ -13,4 +13,18 @@
 		<filename plugin="local">local.php</filename>
 		<folder>adapter</folder>
 	</files>
+
+	<config>
+		<fields name="params">
+			<fieldset name="basic">
+				<field
+						name="display_name"
```
Comment (laoneo) [inline plugins/filesystem/local/local.xml]: I guess there is one tab to much.

File: plugins/filesystem/local/local.xml [hunk 9]
```diff
@@ -13,4 +13,18 @@
...
+						type="text"
+						label="Display Name"
```
Comment (laoneo) [inline plugins/filesystem/local/local.xml]: Needs to be a a translatable key.

File: plugins/filesystem/local/local.xml [hunk 10]
```diff
@@ -13,4 +13,18 @@
...
+						description="Local Filesystem plugin"
```
Comment (laoneo) [inline plugins/filesystem/local/local.xml]: Needs to be a a translatable key.

File: plugins/filesystem/foo/foo.xml [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+	<description>Foo plugin</description>
+	<files>
+		<filename plugin="foo">foo.php</filename>
+		<folder>adapter</folder>
+	</files>
+	<config>
+		<fields name="params">
+			<fieldset name="basic">
+				<field
+						name="display_name"
```
Comment (laoneo) [inline plugins/filesystem/foo/foo.xml]: A tab to much.

File: plugins/filesystem/foo/foo.php [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+	 * Returns a local media adapter to the caller which can be used to manipulate files
+	 *
+	 * @return   MediaFileAdapterFoo
+	 *
+	 * @since    __DEPLOY_VERSION__
+	 */
+	public function onFileSystemGetAdapters()
+	{
+		// Compile the root path
+		$root = JPATH_ROOT . '/fooFiles';
```
Comment (laoneo) [inline plugins/filesystem/foo/foo.php:40]: Should we not ensure that the folder exists?

File: administrator/components/com_media/views/media/tmpl/default.php [hunk 13]
```diff
@@ -11,6 +11,30 @@
...
+{
+	$providerInfo = JPluginHelper::getPlugin('filesystem');
+	$results      = array();
+
+	foreach ($providerInfo as $provider)
+	{
+		$params            = new JRegistry($provider->params);
```
Comment (laoneo) [inline administrator/components/com_media/views/media/tmpl/default.php]: Should be Joomla\Registry\Registry, JRegistry is deprecated.

File: administrator/components/com_media/views/media/tmpl/default.php [hunk 14]
```diff
@@ -11,6 +11,30 @@
...
+/**
+ * Obtain list of supported providers
+ *
+ * @return string
+ *
+ * @since __DEPLOY_VERSION__
+ */
+function getProviderInfo()
+{
+	$providerInfo = JPluginHelper::getPlugin('filesystem');
```
Comment (laoneo) [inline administrator/components/com_media/views/media/tmpl/default.php]: Use the namespaced class.

File: administrator/components/com_media/models/api.php [hunk 15]
```diff
@@ -38,85 +46,126 @@ public function __construct($config = array())
 	{
 		parent::__construct($config);
 
+		if (!isset($config['providers']))
+		{
+			$config['providers'] = JPluginHelper::getPlugin('filesystem');
```
Comment (laoneo) [inline administrator/components/com_media/models/api.php]: Use the namespaced class.

File: plugins/filesystem/foo/foo.php (hunk 12 above)
Comment (kasvith) [reply plugins/filesystem/foo/foo.php:40]: its a mock file system, so i did not put much effort on that

File: plugins/filesystem/foo/foo.xml (hunk 11 above)
Comment (kasvith) [reply plugins/filesystem/foo/foo.xml]: we will remove foo related files soon

File: administrator/components/com_media/controllers/api.json.php [hunk 16]
```diff
@@ -81,7 +81,8 @@ class MediaControllerApi extends Controller
 	public function files()
 	{
 		// Get the required variables
-		$path = $this->input->getPath('path', '/', 'path');
+		list($adapterInfo, $path) = explode(':', $this->input->getString('path', 'local:/'));
+		list($adapter, $account)  = array_merge(explode('-', $adapterInfo), array(""));
```
Comment (dneukirchen) [inline administrator/components/com_media/controllers/api.json.php]: This shows how multiple accounts can be integrated. 
Imo we should not include it until we really use it. Please remove the account handling.

File: administrator/components/com_media/views/media/tmpl/default.php [hunk 17]
```diff
@@ -43,6 +43,8 @@
 	'editViewUrl'             => JUri::root() . 'administrator/index.php?option=com_media&view=file' . $tmpl,
 	'allowedUploadExtensions' => $params->get('upload_extensions', ''),
 	'maxUploadSizeMb'         => $params->get('upload_maxsize', 10),
+	'defaultFileSystem'       => 'local',
```
Comment (dneukirchen) [inline administrator/components/com_media/views/media/tmpl/default.php]: This is hardcoded and should be removed. You can add a "is_default: true/false" to the providers array.

File: administrator/components/com_media/models/api.php [hunk 18]
```diff
@@ -38,85 +46,126 @@ public function __construct($config = array())
 	{
 		parent::__construct($config);
 
+		if (!isset($config['providers']))
+		{
+			$config['providers'] = Joomla\CMS\Plugin\PluginHelper::getPlugin('filesystem');
+			$this->providers = $config['providers'];
```
Comment (dneukirchen) [inline administrator/components/com_media/models/api.php]: What happens when config['providers'] is set?
This should be moved outside of the if block to allow setting providers via config.

File: administrator/components/com_media/models/api.php [hunk 19]
```diff
@@ -38,85 +46,126 @@ public function __construct($config = array())
...
 
-			if ($results != null)
+			for ($i = 0, $len = count($results); $i < $len; $i++)
 			{
-				$config['fileadapter'] = $results[0];
+				$adapters[$this->providers[$i]->name] = $results[$i];
 			}
+
+			$config['fileadapter'] = $adapters;
+			$this->adapters = $adapters;
```
Comment (dneukirchen) [inline administrator/components/com_media/models/api.php]: Same as above: What happens when config['fileadapter'] is set?
This should be moved outside of the if block to allow setting fileadapter via config.

Additionaly the naming should be improved: you use singular config['fileadapter'] and set it to plural $adapters.

File: administrator/components/com_media/models/api.php [hunk 20]
```diff
@@ -38,85 +46,126 @@ public function __construct($config = array())
...
+	private function getAdapter($name)
+	{
+		if (isset($this->adapters[$name]))
 		{
-			$this->adapter = $config['fileadapter'];
+			return $this->adapters[$name];
 		}
+
+		// Todo Use a translated string
+		throw new Exception('Requested media file adapter was not found', 500);
```
Comment (dneukirchen) [inline administrator/components/com_media/models/api.php]: Change to InvalidArgumentException to be more expressive.

**Review by dneukirchen — CHANGES_REQUESTED** [review]

File: administrator/components/com_media/models/api.php [hunk 21]
```diff
@@ -38,85 +46,126 @@ public function __construct($config = array())
 	{
 		parent::__construct($config);
 
+		if (!isset($config['providers']))
+		{
+			$config['providers'] = Joomla\CMS\Plugin\PluginHelper::getPlugin('filesystem');
```
Comment (laoneo) [inline administrator/components/com_media/models/api.php]: This assignment is not needed at all. It makes sense when you want to inject the providers trough the config.

File: administrator/components/com_media/models/api.php [hunk 22]
```diff
@@ -38,85 +46,126 @@ public function __construct($config = array())
...
+			$adapters = array();
 
-			if ($results != null)
+			for ($i = 0, $len = count($results); $i < $len; $i++)
 			{
-				$config['fileadapter'] = $results[0];
+				$adapters[$this->providers[$i]->name] = $results[$i];
 			}
+
+			$config['fileadapter'] = $adapters;
```
Comment (laoneo) [inline administrator/components/com_media/models/api.php]: Same here. This assignment is not needed at all. It makes sense when you want to inject the adapters trough the config.

File: administrator/components/com_media/views/media/view.html.php [hunk 23]
```diff
@@ -18,6 +18,28 @@
...
+	 *
+	 * @param   array  $config  Configuration array
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function __construct(array $config = array())
+	{
+		parent::__construct($config);
+
+		$this->providers = $this->getProviders();
```
Comment (laoneo) [inline administrator/components/com_media/views/media/view.html.php]: No need to do it in the constructor, better in the display function.

File: administrator/components/com_media/models/api.php (hunk 19 above)
Comment (kasvith) [reply administrator/components/com_media/models/api.php]: Sorry for typo

@laoneo do we want to remove this config part from api? It was there in the original MM code when i modified it

File: administrator/components/com_media/views/media/view.html.php [hunk 24]
```diff
@@ -85,4 +108,28 @@ protected function prepareToolbar()
...
 	}
+
+	/**
+	 * Obtain list of supported providers
+	 *
+	 * @return string
+	 *
+	 * @since __DEPLOY_VERSION__
+	 */
+	public function getProviders()
```
Comment (laoneo) [inline administrator/components/com_media/views/media/view.html.php]: This function needs to be moved to the model. Then you can do in this class `$this->providers = $this->get('Providers');`

File: administrator/components/com_media/views/media/view.html.php [hunk 25]
```diff
@@ -85,4 +108,28 @@ protected function prepareToolbar()
...
+		foreach ($providerInfo as $provider)
+		{
+			$params            = new Joomla\Registry\Registry($provider->params);
+			$info              = new stdClass;
+			$info->name        = $provider->name;
+			$info->displayName = $params->get('display_name');
+			$results[]         = $info;
+		}
+
+		return json_encode((array) $results);
```
Comment (laoneo) [inline administrator/components/com_media/views/media/view.html.php]: This needs to be moved to the default.php file.

File: administrator/components/com_media/views/media/view.html.php [hunk 26]
```diff
@@ -34,6 +56,7 @@ public function display($tpl = null)
 		parent::display($tpl);
 	}
 
+
```
Comment (laoneo) [inline administrator/components/com_media/views/media/view.html.php]: Why this extra new line?

File: administrator/components/com_media/views/media/tmpl/default.php (hunk 17 above)
Comment (kasvith) [reply administrator/components/com_media/views/media/tmpl/default.php]: I'll fix that

File: administrator/components/com_media/views/media/view.html.php [hunk 27]
```diff
@@ -18,6 +18,28 @@
...
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function __construct(array $config = array())
+	{
+		parent::__construct($config);
+
+		$this->providers = $this->getProviders();
+	}
+
+
```
Comment (laoneo) [inline administrator/components/com_media/views/media/view.html.php]: One empty line to much.

**Review by laoneo — CHANGES_REQUESTED** [review]

File: administrator/components/com_media/models/api.php (hunk 20 above)
Comment (kasvith) [reply administrator/components/com_media/models/api.php]: Sure

File: administrator/components/com_media/models/api.php (hunk 19 above)
Comment (laoneo) [inline administrator/components/com_media/models/api.php]: It was different in the actual implementation. How I did it was to allow to inject the adapter.

Comment (laoneo) [reply administrator/components/com_media/models/api.php]: For me it makes sense to have a config option, it allows to inject the adapters and providers trough the constructor. But you need to follow the recommendation from @dneukirchen.

File: administrator/components/com_media/models/api.php [hunk 28]
```diff
@@ -38,85 +46,126 @@ public function __construct($config = array())
...
 		}
+	}
+
+	/**
+	 * Returns the enabled adapter list with their properties
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 * @return stdClass[]
+	 */
+	public function getProviders()
```
Comment (laoneo) [inline administrator/components/com_media/models/api.php]: Where is this function needed? Where do we need the providers at all in the api layer?

File: administrator/components/com_media/controllers/api.json.php (hunk 16 above)
Comment (kasvith) [reply administrator/components/com_media/controllers/api.json.php]: this code will try to extract an account info if it presents, we will use this ASAP

Comment (kasvith) [reply administrator/components/com_media/controllers/api.json.php]: we can replace this with `list($adapter) = $adapterInfo` if you want

Comment (laoneo) [reply administrator/components/com_media/controllers/api.json.php]: Better to use only code which is in use actually, when we need it then we can add it at that time.

File: administrator/components/com_media/models/api.php [hunk 29]
```diff
@@ -38,85 +46,116 @@ public function __construct($config = array())
...
 	 *
 	 * @since   __DEPLOY_VERSION__
 	 * @throws  Exception
 	 * @see     MediaFileAdapterInterface::getFile()
 	 */
-	public function getFiles($path = '/', $filter = '')
+	public function getFiles($adapter, $path = '/', $filter = '')
 	{
-		if (!$this->adapter)
+		if (!isset($this->adapters[$adapter]))
```
Comment (laoneo) [inline administrator/components/com_media/models/api.php]: Is this needed? You don't do the check on the other functions?

Comment (laoneo) [inline administrator/components/com_media/models/api.php]: Is this needed? You don't do the check on the other functions?

File: administrator/components/com_media/views/media/tmpl/default.php [hunk 30]
```diff
@@ -43,6 +43,7 @@
 	'editViewUrl'             => JUri::root() . 'administrator/index.php?option=com_media&view=file' . $tmpl,
 	'allowedUploadExtensions' => $params->get('upload_extensions', ''),
 	'maxUploadSizeMb'         => $params->get('upload_maxsize', 10),
+	'providers'               => json_encode((array) $this->get('providers')),
```
Comment (laoneo) [inline administrator/components/com_media/views/media/tmpl/default.php]: Here you have to do `$this->providers`.

File: administrator/language/en-GB/en-GB.plg_filesystem_local.ini [hunk 31]
```diff
@@ -4,4 +4,7 @@
 ; Note : All ini files need to be saved as UTF-8
 
 PLG_FILESYSTEM_LOCAL = "FileSystem - Local"
-PLG_FILESYSTEM_LOCAL_XML_DESCRIPTION = "Local file system plugin for Joomla!"
\ No newline at end of file
+PLG_FILESYSTEM_LOCAL_XML_DESCRIPTION = "Local file system plugin for Joomla!"
+
+PLG_FILESYSTEM_LOCAL_DISPLAY_NAME_LABEL = "Display Name"
+PLG_FILESYSTEM_LOCAL_DISPLAY_NAME_DESC = "Local Filesystem plugin"
```
Comment (laoneo) [inline administrator/language/en-GB/en-GB.plg_filesystem_local.ini]: It needs an empty line at the end of the file.

**Review by laoneo — CHANGES_REQUESTED** [review]

File: administrator/components/com_media/models/api.php (hunk 29 above)
Comment (kasvith) [reply administrator/components/com_media/models/api.php]: yep we need to send an empty array for frontend if adapter is not set

File: administrator/components/com_media/models/api.php [hunk 32]
```diff
@@ -38,85 +46,116 @@ public function __construct($config = array())
...
 
-		if (!isset($config['fileadapter']))
+		if (!isset($config['providers']))
 		{
-			// Import Local file system plugin
-			JPluginHelper::importPlugin('filesystem');
+			$config['providers'] = Joomla\CMS\Plugin\PluginHelper::getPlugin('filesystem');
+		}
+
+		$this->providers = $config['providers'];
```
Comment (laoneo) [inline administrator/components/com_media/models/api.php]: What's the purpose of the providers in this class?

File: administrator/components/com_media/models/media.php [hunk 33]
```diff
@@ -0,0 +1,51 @@
...
+ * @since  __DEPLOY_VERSION__
+ */
+class MediaModelMedia extends Model
+{
+	/**
+	 * @var array|string Holds a list of providers
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	protected $providers = null;
```
Comment (laoneo) [inline administrator/components/com_media/models/media.php]: What do you need that for?

File: administrator/components/com_media/models/api.php (hunk 29 above)
Comment (laoneo) [reply administrator/components/com_media/models/api.php]: Why not throw the exception when the adapter doesn't exist?

File: administrator/components/com_media/models/api.php (hunk 32 above)
Comment (kasvith) [reply administrator/components/com_media/models/api.php]: we will need it for multiple accounts soon

File: administrator/components/com_media/models/media.php (hunk 33 above)
Comment (kasvith) [reply administrator/components/com_media/models/media.php]: for multiple accounts we need it

File: administrator/language/en-GB/en-GB.plg_filesystem_local.ini (hunk 31 above)
Comment (kasvith) [reply administrator/language/en-GB/en-GB.plg_filesystem_local.ini]: done

File: administrator/components/com_media/models/media.php (hunk 33 above)
Comment (laoneo) [reply administrator/components/com_media/models/media.php]: I don't see a place where this variable is used.

File: administrator/language/en-GB/en-GB.plg_filesystem_local.ini [hunk 34]
```diff
@@ -1,7 +1 @@
...
+; Joomla! Project
; Copyright (C) 2005 - 2017 Open Source Matters. All rights reserved.
; License GNU General Public License version 2 or later; see LICENSE.txt, see LICENSE.php
; Note : All ini files need to be saved as UTF-8

PLG_FILESYSTEM_LOCAL = "FileSystem - Local"
PLG_FILESYSTEM_LOCAL_XML_DESCRIPTION = "Local file system plugin for Joomla!"

PLG_FILESYSTEM_LOCAL_DISPLAY_NAME_LABEL = "Display Name"
PLG_FILESYSTEM_LOCAL_DISPLAY_NAME_DESC = "Local Filesystem plugin"
```
Comment (laoneo) [inline administrator/language/en-GB/en-GB.plg_filesystem_local.ini]: Guess you screwed it up here

File: administrator/components/com_media/models/api.php [hunk 35]
```diff
@@ -19,12 +19,20 @@
...
-	protected $adapter = null;
+	protected $adapters = null;
+
+	/**
+	 * Stores providers for filesystem plugin group
+	 *
+	 * @var    stdClass[]
+	 * @since  __DEPLOY_VERSION__
+	 */
+	protected $providers = null;
```
Comment (laoneo) [inline administrator/components/com_media/models/api.php]: Where is this variable used?

File: administrator/language/en-GB/en-GB.plg_filesystem_local.ini (hunk 34 above)
Comment (kasvith) [reply administrator/language/en-GB/en-GB.plg_filesystem_local.ini]: seems crlf was used, fixed asap

File: administrator/components/com_media/models/api.php (hunk 35 above)
Comment (kasvith) [reply administrator/components/com_media/models/api.php]: not used at the moment, for multiple accounts i decided to use it

File: administrator/components/com_media/models/media.php (hunk 33 above)
Comment (kasvith) [reply administrator/components/com_media/models/media.php]: yep ATM it has not used, i'll remove it

**Comment by dneukirchen** [comment]
good job @kasvith

**Comment by kasvith** [comment]
thank you @dneukirchen

**Pull request merged**